File: flutter_inappwebview_forge_android/android/src/main/kotlin/com/emirkanacar/flutter_inappwebview_forge_android/headless_in_app_webview/HeadlessInAppWebViewManager.kt
package com.emirkanacar.flutter_inappwebview_forge_android.headless_in_app_webview

import android.content.Context
import com.emirkanacar.flutter_inappwebview_forge_android.InAppWebViewFlutterPlugin
import com.emirkanacar.flutter_inappwebview_forge_android.types.ChannelDelegateImpl
import com.emirkanacar.flutter_inappwebview_forge_android.webview.in_app_webview.FlutterWebView
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.HashMap

open class HeadlessInAppWebViewManager(
    initialPlugin: InAppWebViewFlutterPlugin
) : ChannelDelegateImpl(
    MethodChannel(initialPlugin.requireMessenger(), METHOD_CHANNEL_NAME)
) {
    companion object {
        @JvmField
        protected val LOG_TAG = "HeadlessInAppWebViewManager"

        @JvmField
        val METHOD_CHANNEL_NAME = "com.emirkanacar/flutter_headless_inappwebview"
    }

    @JvmField
    val webViews: MutableMap<String, HeadlessInAppWebView?> = HashMap()

    @JvmField
    var plugin: InAppWebViewFlutterPlugin? = initialPlugin

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val id = call.argument<String>("id")
        if (id == null) {
            result.error("invalid_arguments", "id is required.", null)
            return
        }

        when (call.method) {
            "run" -> {
                val params = call.argument<HashMap<String, Any?>>("params") ?: hashMapOf()
                run(id, params)
                result.success(true)
            }
            else -> result.notImplemented()
        }
    }

    fun run(id: String, params: HashMap<String, Any?>) {
        val currentPlugin = plugin ?: return
        val context: Context = currentPlugin.activity ?: currentPlugin.applicationContext ?: return
        val deferNativeOperationsUntilAttach = currentPlugin.activity != null
        val flutterWebView = FlutterWebView(
            currentPlugin,
            context,
            id,
            params,
            false
        )
        val headlessInAppWebView = HeadlessInAppWebView(currentPlugin, id, flutterWebView)
        webViews[id] = headlessInAppWebView

        headlessInAppWebView.prepare(params)
        headlessInAppWebView.onWebViewCreated()
        flutterWebView.makeInitialLoad(params, deferNativeOperationsUntilAttach)
    }

    override fun dispose() {
        super.dispose()
        webViews.values.forEach { it?.dispose() }
        webViews.clear()
        plugin = null
    }
}

File: flutter_inappwebview_forge_ios/ios/flutter_inappwebview_forge_ios/Sources/flutter_inappwebview_forge_ios/HeadlessInAppWebView/HeadlessInAppWebViewManager.swift
//
//  HeadlessInAppWebViewManager.swift
//  flutter_inappwebview
//
//  Created by Lorenzo Pichilli on 10/05/2020.
//

import Foundation

import Flutter
import UIKit
import WebKit
import Foundation
import AVFoundation

public class HeadlessInAppWebViewManager: ChannelDelegate {
    static let METHOD_CHANNEL_NAME = "com.emirkanacar/flutter_headless_inappwebview"
    var plugin: InAppWebViewFlutterPlugin?
    var webViews: [String: HeadlessInAppWebView] = [:]
    
    init(plugin: InAppWebViewFlutterPlugin) {
        super.init(channel: FlutterMethodChannel(name: HeadlessInAppWebViewManager.METHOD_CHANNEL_NAME, binaryMessenger: plugin.registrar.messenger()))
        self.plugin = plugin
    }
    
    public override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? NSDictionary
        let id: String = arguments!["id"] as! String

        switch call.method {
            case "run":
                let params = arguments!["params"] as! [String: Any?]
                run(id: id, params: params)
                result(true)
                break
            default:
                result(FlutterMethodNotImplemented)
                break
        }
    }
    
    public func run(id: String, params: [String: Any?]) {
        guard let plugin = plugin else {
            return
        }
        webViews.removeValue(forKey: id)?.dispose()
        // Headless and regular WebViews share the native id namespace. Remove
        // and dispose the active owner before registering the replacement so a
        // duplicate id cannot leave a stale channel or WebView reachable.
        if let previousWebView = plugin.inAppWebViewManager?.webViews.removeValue(forKey: id) {
            previousWebView.dispose()
        }
        let flutterWebView = FlutterWebViewController(plugin: plugin,
            withFrame: CGRect.zero,
            viewIdentifier: id,
            params: params as NSDictionary)
        let headlessInAppWebView = HeadlessInAppWebView(plugin: plugin, id: id, flutterWebView: flutterWebView)
        webViews[id] = headlessInAppWebView
        
        headlessInAppWebView.prepare(params: params as NSDictionary)
        headlessInAppWebView.onWebViewCreated()
        flutterWebView.makeInitialLoad(params: params as NSDictionary)
    }
    
    public override func dispose() {
        super.dispose()
        let headlessWebViews = Array(webViews.values)
        webViews.removeAll()
        headlessWebViews.forEach { $0.dispose() }
        plugin = nil
    }
    
    deinit {
        dispose()
    }
}
